executor/common_kvm_arm64_syzos.h

```diff
@@ -0,0 +1,83 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// This file provides guest code running inside the ARM64 KVM.
+
+#include "kvm.h"
+
+// Host will map the code in this section into the guest address space.
+#define GUEST_CODE __attribute__((section("guest")))
+
+// Start/end of the guest section.
+extern char *__start_guest, *__stop_guest;
+
+typedef enum {
+	SYZOS_API_UEXIT,
+	SYZOS_API_CODE,
+	SYZOS_API_STOP, // Must be the last one
+} syzos_api_id;
+
+struct api_call_header {
+	uint64 call;
+	uint64 size;
+};
+
+struct api_call_uexit {
+	struct api_call_header header;
+	uint64 exit_code;
+};
+
+struct api_call_code {
+	struct api_call_header header;
+	uint32 insns[];
+};
+
+void guest_uexit(uint64 exit_code);
+void guest_execute_code(uint32* insns, uint64 size);
+
+GUEST_CODE void interpret_payload(uint64 addr, uint64 size)
+{
+	do {
+		struct api_call_header* cmd = (struct api_call_header*)addr;
+		if (cmd->call >= SYZOS_API_STOP)
+			return;
+		if (cmd->size > size)
+			return;
+		switch (cmd->call) {
+		case SYZOS_API_UEXIT: {
+			struct api_call_uexit* ucmd = (struct api_call_uexit*)cmd;
+			guest_uexit(ucmd->exit_code);
+			break;
+		}
+		case SYZOS_API_CODE: {
+			struct api_call_code* ccmd = (struct api_call_code*)cmd;
+			guest_execute_code(ccmd->insns, cmd->size - sizeof(struct api_call_header));
+			break;
+		}
+		}
+		addr += cmd->size;
+		size -= cmd->size;
+	} while (size > 0);
+}
+
+// Main guest function that performs necessary setup and passes the control to the user-provided
+// payload.
+GUEST_CODE static void guest_main(uint64 size)
+{
+	interpret_payload(ARM64_ADDR_USER_CODE, size);
```

Comment:
I would inline it here.
All this code appears in reproducers, so we balance between code structuring and verbosity.